File: passenger/src/main/java/com/modsen/passenger/service/impl/PassengerServiceImpl.java
package com.modsen.passenger.service.impl;

import com.modsen.passenger.dto.request.PassengerCreate;
import com.modsen.passenger.dto.request.PassengerUpdate;
import com.modsen.passenger.dto.response.PassengerResponse;
import com.modsen.passenger.exception.EmailAlreadyExistsException;
import com.modsen.passenger.exception.PassengerNotFoundException;
import com.modsen.passenger.exception.PhoneNumberAlreadyExistsException;
import com.modsen.passenger.mapper.PassengerMapper;
import com.modsen.passenger.model.Passenger;
import com.modsen.passenger.repository.PassengerRepository;
import com.modsen.passenger.service.PassengerService;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional
@RequiredArgsConstructor
public class PassengerServiceImpl implements PassengerService {

    private final PassengerMapper mapper;
    private final PassengerRepository passengerRepository;

    @Override
    @Transactional(readOnly = true)
    public PassengerResponse findById(Integer id) {
        return passengerRepository.findById(id)
                .map(mapper::toResponse)
                .orElseThrow(() -> new PassengerNotFoundException("exception.passenger_not_found", id));
    }

    @Override
    @Transactional(readOnly = true)
    public Page<PassengerResponse> findAll(Pageable pageable) {
        return passengerRepository.findAll(pageable)
                .map(mapper::toResponse);
    }

    @Override
    public PassengerResponse update(Integer id, PassengerUpdate request) {
        Passenger passenger = passengerRepository.findById(id)
                .orElseThrow(() -> new PassengerNotFoundException("exception.passenger_not_found", id));
        throwIfEmailOrPhoneAlreadyExist(request.getEmail(), request.getPhoneNumber());
        mapper.mapIfNotNull(request, passenger);
        passengerRepository.save(passenger);
        return mapper.toResponse(passenger);
    }


    @Override
    public void deleteById(Integer id) {
        passengerRepository.deleteById(id);
    }

    @Override
    public PassengerResponse create(PassengerCreate request) {
        Passenger passenger = mapper.toModel(request);
        throwIfEmailOrPhoneAlreadyExist(request.getEmail(), request.getPhoneNumber());
        passengerRepository.save(passenger);
        return mapper.toResponse(passenger);
    }

    private void throwIfEmailOrPhoneAlreadyExist(String email, String phoneNumber) {
        passengerRepository.findByEmail(email)
                .ifPresent(ignored -> {
                    throw new EmailAlreadyExistsException("exception.email_already_exists", email);
                });
        passengerRepository.findByPhoneNumber(phoneNumber)
                .ifPresent(ignored -> {
                    throw new PhoneNumberAlreadyExistsException("exception.phone_already_exists", phoneNumber);
                });
    }
}

File: kt-passenger/src/main/kotlin/com/modsen/ktpassenger/service/impl/PassengerServiceImpl.kt
package com.modsen.ktpassenger.service.impl

import com.modsen.ktpassenger.dto.PassengerCreate
import com.modsen.ktpassenger.dto.PassengerResponse
import com.modsen.ktpassenger.dto.PassengerUpdate
import com.modsen.ktpassenger.dto.mapIfNotNull
import com.modsen.ktpassenger.dto.toEntity
import com.modsen.ktpassenger.dto.toResponse
import com.modsen.ktpassenger.exception.EmailAlreadyExistsException
import com.modsen.ktpassenger.exception.PassengerNotFoundException
import com.modsen.ktpassenger.exception.PhoneNumberAlreadyExistsException
import com.modsen.ktpassenger.repository.PassengerRepository
import com.modsen.ktpassenger.service.PassengerService
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class PassengerServiceImpl(
    private val passengerRepository: PassengerRepository
) : PassengerService {

    @Transactional(readOnly = true)
    override fun findById(id: Int): PassengerResponse {
        return passengerRepository.findById(id)
            .map { p -> p.toResponse() }
            .orElseThrow { PassengerNotFoundException("exception.passenger_not_found", id) }
    }

    @Transactional(readOnly = true)
    override fun findAll(pageable: Pageable): Page<PassengerResponse> {
        return passengerRepository.findAll(pageable)
            .map { p -> p.toResponse() }
    }

    override fun update(id: Int, request: PassengerUpdate): PassengerResponse {
        val entity = passengerRepository.findById(id)
            .orElseThrow { PassengerNotFoundException("exception.passenger_not_found", id) }
        throwIfEmailOrPhoneAlreadyExist(request.email ?: "", request.phoneNumber ?: "")
        entity.mapIfNotNull(request)
        passengerRepository.save(entity)
        return entity.toResponse()
    }

    override fun deleteById(id: Int) {
        return passengerRepository.deleteById(id)
    }

    override fun create(request: PassengerCreate): PassengerResponse {
        val entity = request.toEntity()
        throwIfEmailOrPhoneAlreadyExist(entity.email, entity.phoneNumber)
        passengerRepository.save(entity)
        return entity.toResponse()
    }

    private fun throwIfEmailOrPhoneAlreadyExist(email: String, phoneNumber: String) {
        passengerRepository.findByEmail(email)
            .ifPresent { throw EmailAlreadyExistsException("exception.email_already_exists", email) }
        passengerRepository.findByPhoneNumber(phoneNumber)
            .ifPresent { throw PhoneNumberAlreadyExistsException("exception.phone_already_exists", phoneNumber) }
    }
}